data_generators/xcspec_extractor.py

```diff
@@ -63,29 +65,147 @@ def xcspec_from_file(path):
+        return 'return (False, %s)' % repr(string)
+
+    expanded_variables = {}
+
+    def record_substitution(match):
+        name = match[1]
+        if name not in expanded_variables:
+            expanded_variables[name] = (
+                'eval_val_%s' % len(expanded_variables),
+                name,
+                'eval_key_%s' % len(expanded_variables),
+            )
+        if key == "DefaultValue":
+            return "{%s}" % expanded_variables[name][0]
+        return expanded_variables[name][0]
+
+    while True:
+        (string, subs_made) = XCCONFIG_EXPANSION.subn(
+            repl=record_substitution, string=string)
+        if not subs_made:
+            break
+
+    def format_str(s):
+        val_references = STRING_FORMAT_EVAL_VAL_REFERENCE.findall(s)
+        if not val_references:
+            return '"%s"' % s
+        if len(val_references) == 1:
+            if s == '{%s}' % val_references[0]:
+                # directly return the variable, without using an format string
+                return val_references[0]
+
+        return '"{string}".format({format})'.format(
+            string=s,
+            format=', '.join(('{var} = {var}'.format(var=v) for v in val_references)))
+
+    method_body = "\n    used_user_content = False\n"
+    method_body += "\n".join((
+        """
+    {var} = ""
+    {key} = {key_value}
+    if {key} in xcconfigs:
+        {var} = xcconfigs[{key}]
+        used_user_content = True
+    elif {key} in id_configs:
+        opt = id_configs[{key}]
+        if "DefaultValue" in opt:
+            ({var}_used_user_content, {var}) = XCSPEC_EVALS[opt["DefaultValue"]](xcconfigs, id_configs)
+            used_user_content = used_user_content or {var}_used_user_content""".format(var=var, name=name, key=key, key_value=format_str(name))
+        for (name, (var, _i, key)) in expanded_variables.items()
+    ))
+
+    if key == 'DefaultValue':
+        return method_body + '\n\n    return (used_user_content, %s)' % (format_str(string), )
+    elif key == 'Condition':
+        string = CONDITION_EVAL_VAL_REFERENCE.sub(
+            repl=lambda m: repr(m[1]), string=string)
+        return method_body + \
+            '\n\n    return (used_user_content, ({}))'.format(
+                string.replace('||', 'or').replace('!=', '__NEQ__').replace(
+                    '!', 'not ').replace('&&', 'and').replace('__NEQ__', '!=')
+            )
+
+    raise "Unknown key %s" % key
+
+
+def add_eval(id, name, key, string, eval_methods):
+    method_name = '__'.join(
+        (id.replace('.', '_'), name.replace('.', '_'), key))
+    if method_name in eval_methods:
+        raise "Dup"
+
+    eval_methods[method_name] = """
+def _{method_name}(xcconfigs, id_configs):
+    # {string}
+    {expr}
+""".format(method_name=method_name, string=string, expr=generate_method_body_for_expression(string, key))
+
+    return method_name
 
-SETTINGS = """ % xcode_version
 
 to_print = {
```

Comment:
Not a biggie and we can fix it later but I tend to find python scripts without a `main` and with code scattered all around the global namespace hard to read. I'd take all the logic and shove it in a `main()` and do the standard:

```
if __name__ == '__main__':
    main()
````

dance.